bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/TypeChecker.java

```diff
@@ -1246,8 +1246,62 @@ private static boolean hasIncompatibleReadOnlyFlags(Field targetField, Field sou
                                                                                          Flags.READONLY);
     }
 
+    private static boolean checkIsArrayType(BArrayType sourceType, BArrayType targetType,
+                                            List<TypePair> unresolvedTypes) {
+        Type sourceElementType = sourceType.getElementType();
+        Type targetElementType = targetType.getElementType();
+
+        switch (sourceType.getState()) {
+            case OPEN:
+                if (targetType.getState() != ArrayState.OPEN) {
+                    return false;
+                }
+                break;
+            case CLOSED:
+                if (targetType.getState() == ArrayState.CLOSED &&
+                        sourceType.getSize() != targetType.getSize()) {
+                    return false;
+                }
+                break;
+        }
+        return checkIsType(sourceElementType, targetElementType, unresolvedTypes);
+    }
+
+    private static boolean checkIsArrayType(BTupleType sourceType, BArrayType targetType,
+                                            List<TypePair> unresolvedTypes) {
+        Set<Type> tupleTypes = new HashSet<>(sourceType.getTupleTypes());
```

Comment:
Do we need a new set here? Can't we just work with the list returned by `sourceType.getTupleTypes()`?